File: packages/appNuxt/database/data/Restaurants.ts
// import { GeoHashUtils } from '@app/tools'
import { IFBRestaurant } from 'ieattatypes'
import { slugifyToLower } from '~/database/slug_helper'

export const loadRestaurants = (): IFBRestaurant[] => {
  const next = restaurants.map((item: IFBRestaurant) => {
    // const geoHash = GeoHashUtils.encode(item.latitude, item.longitude)
    // item.geoHash = geoHash
    item.slug = slugifyToLower(item.displayName)
    // console.log(JSON.stringify(item))
    return item
  })
  // console.log(JSON.stringify(next))
  return next
}

export const restaurants: IFBRestaurant[] =
  [{
    isNew: false,
    uniqueId: '035ac47c-5781-4da8-af21-35c97a46c101',
    extraNote: '',
    creatorId: '',
    flag: '1',
    address: '724 S Broadway, Los Angeles, CA 90014, USA',
    displayName: 'Forno Vecchio',
    originalUrl: 'http://res.cloudinary.com/di3fvexj8/image/upload/v1507529261/politicl/o_xr3usf.jpg',
    thumbnailUrl: 'http://res.cloudinary.com/di3fvexj8/image/upload/s--qsTnagxL--/c_fill,f_auto,h_348,q_auto,w_348/v1/politicl/o_xr3usf',
    geoHash: 'hb00n8n2pb08',
    latitude: 34.044561,
    longitude: -118.253985,
    createdAt: '2017-10-09T06:02:09.838+0000',
    updatedAt: '2017-11-07T07:43:10.690+0000',
    rate: 3.5,
    reviewCount: 12,
    street_number: '724',
    route: 'South Broadway',
    locality: 'Los Angeles',
    sublocality: '',
    country: 'US',
    postal_code: '2802',
    administrative_area: 'CA',
    slug: 'forno-vecchio'
  }, {
    isNew: false,
    uniqueId: '0b92b483-8860-438b-961d-4fef4b124176',
    extraNote: '',
    creatorId: '',
    flag: '1',
    address: '250s S Broadway, Los Angeles, CA 90012, USA',
    displayName: 'Carl\'s Jr',
    originalUrl: 'http://res.cloudinary.com/di3fvexj8/image/upload/v1507529311/politicl/o_ugrhuw.jpg',
    thumbnailUrl: 'http://res.cloudinary.com/di3fvexj8/image/upload/s--TqIOho75--/c_fill,f_auto,h_348,q_auto,w_348/v1/politicl/o_ugrhuw',
    geoHash: 'hb00nb521bn8',
    latitude: 34.051178,
    longitude: -118.247636,
    createdAt: '2017-10-09T06:02:52.325+0000',
    updatedAt: '2017-10-09T06:35:46.537+0000',
    rate: 2.5,
    reviewCount: 3,
    street_number: '250s',
    route: 'South Broadway',
    locality: 'Los Angeles',
    sublocality: '',
    country: 'US',
    postal_code: '3605',
    administrative_area: 'CA',
    slug: 'carl\'s-jr'
  }, {
    isNew: false,
    uniqueId: 'f603f752-5641-472d-bcd1-0dec921b8931',
    extraNote: '',
    creatorId: '',
    flag: '1',
    address: '336 S Hill St, Los Angeles, CA 90013, USA',
    displayName: 'La Cita',
    originalUrl: 'http://res.cloudinary.com/di3fvexj8/image/upload/v1507529351/politicl/o_h1fei1.jpg',
    thumbnailUrl: 'http://res.cloudinary.com/di3fvexj8/image/upload/s--r05hgf5R--/c_fill,f_auto,h_348,q_auto,w_348/v1/politicl/o_h1fei1',
    geoHash: 'hb00nb521bn8',
    latitude: 34.050796,
    longitude: -118.249782,
    createdAt: '2017-10-09T06:03:49.167+0000',
    updatedAt: '2017-10-09T06:36:46.436+0000',
    rate: 2.0,
    reviewCount: 15,
    street_number: '336',
    route: 'South Hill Street',
    locality: 'Los Angeles',
    sublocality: '',
    country: 'US',
    postal_code: '1109',
    administrative_area: 'CA',
    slug: 'la-cita'
  }, {
    isNew: false,
    uniqueId: '0423ce1f-34c5-405a-9b7a-22202046ee68',
    extraNote: '',
    creatorId: '',
    flag: '1',
    address: '401 S Grand Ave, Los Angeles, CA 90071, USA',
    displayName: 'Pez Cantina',
    originalUrl: 'http://res.cloudinary.com/di3fvexj8/image/upload/v1507529398/politicl/o_npi8ev.jpg',
    thumbnailUrl: 'http://res.cloudinary.com/di3fvexj8/image/upload/s--pdqw12x3--/c_fill,f_auto,h_348,q_auto,w_348/v1/politicl/o_npi8ev',
    geoHash: 'hb00nb584012',
    latitude: 34.051525,
    longitude: -118.253312,
    createdAt: '2017-10-09T06:04:35.361+0000',
    updatedAt: '2017-10-09T06:37:45.146+0000',
    rate: 4.0,
    reviewCount: 8,
    street_number: '401',
    route: 'South Grand Avenue',
    locality: 'Los Angeles',
    sublocality: '',
    country: 'US',
    postal_code: '90071',
    administrative_area: 'CA',
    slug: 'pez-cantina'
  }, {
    isNew: false,
    uniqueId: 'e854bd77-7836-4244-8b17-f7b2ec5e910a',
    extraNote: '',
    creatorId: '',
    flag: '1',
    address: '517 W 6th St, Los Angeles, CA 90014, USA',
    displayName: 'Water Grill',
    originalUrl: 'http://res.cloudinary.com/di3fvexj8/image/upload/v1507529446/politicl/o_ygnkam.jpg',
    thumbnailUrl: 'http://res.cloudinary.com/di3fvexj8/image/upload/s--sX0P-dsK--/c_fill,f_auto,h_348,q_auto,w_348/v1/politicl/o_ygnkam',
    geoHash: 'hb00nb1b18n2',
    latitude: 34.049107,
    longitude: -118.254707,
    createdAt: '2017-10-09T06:05:13.322+0000',
    updatedAt: '2017-10-09T06:39:17.107+0000',
    rate: 1.0,
    reviewCount: 20,
    street_number: '517',
    route: 'West 6th Street',
    locality: 'Los Angeles',
    sublocality: '',
    country: 'US',
    postal_code: '1201',
    administrative_area: 'CA',
    slug: 'water-grill'
  }, {
    isNew: false,
    uniqueId: 'f1c0aff9-728b-4041-9560-c09578ce7b01',
    extraNote: '',
    creatorId: '',
    flag: '1',
    geoHash: 'uxzpcrfzypbp',
    latitude: 32.402716,
    longitude: 120.550209,
    displayName: 'trujunzhang locally',
    createdAt: '2017-10-10T06:05:35.316+0000',
    updatedAt: '2017-10-10T06:05:35.633+0000',
    rate: 1.5,
    reviewCount: 4,
    thumbnailUrl: 'http://res.cloudinary.com/di3fvexj8/image/upload/s--zE0P3i_b--/c_fill,f_auto,h_348,q_auto,w_348/v1/politicl/b2eccd155ad6639c29e51f9b0d2549f6_image_bjvqag',
    originalUrl: 'http://res.cloudinary.com/di3fvexj8/image/upload/v1507615599/politicl/b2eccd155ad6639c29e51f9b0d2549f6_image_bjvqag.jpg',
    address: '399 Ren Shou Lu, Rugao Shi, Nantong Shi, Jiangsu Sheng, China, 226500',
    street_number: '399',
    route: 'Ren Shou Lu',
    locality: 'Nantong Shi',
    sublocality: 'Rugao Shi',
    country: 'CN',
    postal_code: '226500',
    administrative_area: 'Jiangsu Sheng',
    slug: 'trujunzhang-locally'
  }]

File: packages/appGetx/plugin/getx_firebase/lib/src/data/model/Restaurants.dart
import 'package:my_plugin/my_plugin.dart';
import '../../utils/geohash_utils.dart';
import '../../utils/slug_helper.dart';
import '../../utils/timeago_utils.dart';

import 'Base_Review.dart';
import 'Database.dart';
import 'auth_user_model.dart';

class ParseModelRestaurants extends BaseReview {
  // Base(5)
  final String uniqueId;
  final String creatorId;
  final String createdAt;
  String updatedAt;
  final String flag;

  // extra(1)
  String extraNote;

  // Check google(1)
  final bool isNew;

  // Location(3)
  final String geoHash;
  final double latitude;
  final double longitude;

  // Common(4)
  String displayName;
  String slug;
  final String thumbnailUrl;
  String originalUrl;

  // for review(2)
  int rate;
  int reviewCount;

  // Google api(8)
  final String? address;
  final String? street_number;
  final String? route;
  final String? locality;
  final String? sublocality;
  final String? country;
  final String? postal_code;
  final String? administrative_area;

  ParseModelRestaurants({
    // Base(5)
    required this.uniqueId,
    required this.creatorId,
    required this.createdAt,
    required this.updatedAt,
    required this.flag,
    // extra(1)
    required this.extraNote,
    // Check google(1)
    required this.isNew,
    // Location(3)
    required this.geoHash,
    required this.latitude,
    required this.longitude,
    // Common(4)
    required this.displayName,
    required this.slug,
    required this.thumbnailUrl,
    required this.originalUrl,
    // for review(2)
    required this.rate,
    required this.reviewCount,
    // Google api(8)
    required this.address,
    required this.street_number,
    required this.route,
    required this.locality,
    required this.sublocality,
    required this.country,
    required this.postal_code,
    required this.administrative_area,
  }) : super(rate, reviewCount);

  static emptyRestaurant({
    required AuthUserModel? authUserModel,
    required double latitude,
    required double longitude,
  }) {
    return ParseModelRestaurants(
      // Base(5)
      uniqueId: documentIdFromCurrentDate(),
      creatorId: authUserModel!.uid,
      createdAt: getDateStringForCreatedOrUpdatedDate(),
      updatedAt: getDateStringForCreatedOrUpdatedDate(),
      flag: '1',
      // extra(1)
      extraNote: '',
      // Check google(1)
      isNew: true,
      // Location(3)
      geoHash: convertToGeoHash(latitude, longitude),
      latitude: latitude,
      longitude: longitude,
      // Common(4)
      displayName: '',
      slug: '',
      thumbnailUrl: '',
      originalUrl: '',
      // for review(2)
      rate: 0,
      reviewCount: 0,
      // Google api(8)
      address: '',
      street_number: '',
      route: '',
      locality: '',
      sublocality: '',
      country: '',
      postal_code: '',
      administrative_area: '',
    );
  }

  static ParseModelRestaurants updateCover(
      {required ParseModelRestaurants model, required String originalUrl}) {
    model.originalUrl = originalUrl;
    model.updatedAt = getDateStringForCreatedOrUpdatedDate();

    return model;
  }

  static ParseModelRestaurants updateRestaurant(
      {required ParseModelRestaurants model,
      required String nextDisplayName,
      required String nextExtraNote}) {
    model.displayName = nextDisplayName;
    model.slug = slugifyToLower(nextDisplayName);
    model.extraNote = nextExtraNote;
    model.updatedAt = getDateStringForCreatedOrUpdatedDate();

    return model;
  }

  factory ParseModelRestaurants.fromJson(Map<String, dynamic> json) {
    // Base(5)
    DatabaseBaseModel databaseBaseModel = DatabaseBaseModel.fromJson(json);

    // extra(1)
    String extraNote = json['extraNote'];
    // Check google(1)
    bool isNew = json['isNew'];
    // Location(3)
    String geoHash = json['geoHash'];
    double latitude = json['latitude'];
    double longitude = json['longitude'];
    // Common(4)
    String displayName = json['displayName'];
    String slug = json['slug'];
    String thumbnailUrl = json['thumbnailUrl'];
    String originalUrl = json['originalUrl'];
    // for review(2)
    var rate = json['rate'];
    if (rate is int) {
      rate = rate as int;
    }

    if (rate is double) {
      rate = rate.round();
    }

    int reviewCount = json['reviewCount'];
    // Google api(8)
    String address = json['address'];
    String street_number = json['street_number'];
    String route = json['route'];
    String locality = json['locality'];
    String sublocality = json['sublocality'];
    String country = json['country'];
    String postal_code = json['postal_code'];
    String administrative_area = json['administrative_area'];

    return ParseModelRestaurants(
      // Base(5)
      uniqueId: databaseBaseModel.uniqueId,
      creatorId: databaseBaseModel.creatorId,
      createdAt: databaseBaseModel.createdAt,
      updatedAt: databaseBaseModel.updatedAt,
      flag: databaseBaseModel.flag,
      // extra(1)
      extraNote: extraNote,
      // Check google(1)
      isNew: isNew,
      // Location(3)
      geoHash: geoHash,
      latitude: latitude,
      longitude: longitude,
      // Common(4)
      displayName: displayName,
      slug: slug,
      thumbnailUrl: thumbnailUrl,
      originalUrl: originalUrl,
      // for review(2)
      rate: rate,
      reviewCount: reviewCount,
      // Google api(8)
      address: address,
      street_number: street_number,
      route: route,
      locality: locality,
      sublocality: sublocality,
      country: country,
      postal_code: postal_code,
      administrative_area: administrative_area,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      // Base(5)
      'uniqueId': uniqueId,
      'creatorId': creatorId,
      'createdAt': createdAt,
      'updatedAt': updatedAt,
      'flag': flag,
      // extra(1)
      'extraNote': extraNote,
      // Check google(1)
      'isNew': isNew,
      // Location(3)
      'geoHash': geoHash,
      'latitude': latitude,
      'longitude': longitude,
      // Common(4)
      'displayName': displayName,
      'slug': slug,
      'thumbnailUrl': thumbnailUrl,
      'originalUrl': originalUrl,
      // for review(2)
      'rate': rate,
      'reviewCount': reviewCount,
      // Google api(8)
      'address': address,
      'street_number': street_number,
      'route': route,
      'locality': locality,
      'sublocality': sublocality,
      'country': country,
      'postal_code': postal_code,
      'administrative_area': administrative_area,
    };
  }
}
